### ISSUE TITLE
cargo fix aborts after an error

### ISSUE THREAD

**Issue by mrpink76** [rust-lang/rust#93916]
When i tried to run ```cargo fix``` I got an error message that I should submit a bug with the output of cargo fix which is what I'm doing

`rustc --version --verbose`:
```
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: aarch64-apple-darwin
release: 1.58.1
LLVM version: 13.0.0
```

The error I got after applying ```cargo fix --broken-code```
```
   Compiling rc v0.1.0 (/Users/petter/git/rc)
error[E0425]: cannot find value `input` in this scope
   --> src/filesystem/mod.rs:115:21
    |
115 |                     input.clone(),
    |                     ^^^^^ help: a local variable with a similar name exists: `_input`

error[E0425]: cannot find value `input` in this scope
   --> src/filesystem/mod.rs:118:18
    |
118 |             (id, input)
    |                  ^^^^^ help: a local variable with a similar name exists: `_input`
```

The problematic code:

```rust
    async fn run_background_task(&self, task: Box<dyn BackgroundTask>) -> () {
        let (_id, _task) = {
            let mut app = self.app.lock().await;
            let id = Uuid::new_v4();
            app.foreground_task = Some(id);
            let input = task.task();  // <--- 'input' here is flagged with a warning saying that it is never used (which is wrong)
            app.tasks.insert(
                id,
                (
                    TaskData {
                        start: todo!(),
                        completed: todo!(),
                    },
                    input.clone(),
                ),
            );
            (id, input)
        };
   }
```

The warning:
```
warning: unused variable: `input`
   --> src/filesystem/mod.rs:107:17
    |
107 |             let input = task.task();
    |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_input`
    |
    = note: `#[warn(unused_variables)]` on by default
```

The output of cargo fix:

```
petter@petters-mbp rc % cargo fix
    Checking rc v0.1.0 (/Users/petter/git/rc)
warning: failed to automatically apply fixes suggested by rustc to crate `rc`

after fixes were automatically applied the compiler reported errors within these files:

  * /Users/petter/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/panic.rs
  * src/filesystem/mod.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `input` in this scope
   --> src/filesystem/mod.rs:115:21
    |
115 |                     input.clone(),
    |                     ^^^^^ help: a local variable with a similar name exists: `_input`

error[E0425]: cannot find value `input` in this scope
   --> src/filesystem/mod.rs:118:18
    |
118 |             (id, input)
    |                  ^^^^^ help: a local variable with a similar name exists: `_input`

warning: unreachable expression
   --> src/filesystem/mod.rs:113:36
    |
112 |                         start: todo!(),
    |                                ------- any code following this expression is unreachable
113 |                         completed: todo!(),
    |                                    ^^^^^^^ unreachable expression
    |
    = note: `#[warn(unreachable_code)]` on by default
    = note: this warning originates in the macro `$crate::panic::panic_2015` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: unused import: `crate::app::Pane`
 --> src/filesystem/mod.rs:2:5
  |
2 | use crate::app::Pane;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `crate::app::RcTask::DeleteInput`
 --> src/filesystem/mod.rs:3:5
  |
3 | use crate::app::RcTask::DeleteInput;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `crate::app::RcTask::SearchInput`
 --> src/filesystem/mod.rs:4:5
  |
4 | use crate::app::RcTask::SearchInput;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `crate::filesystem::flatsearchresultprovider::FlatSearchResultProvider`
 --> src/filesystem/mod.rs:8:5
  |
8 | use crate::filesystem::flatsearchresultprovider::FlatSearchResultProvider;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `chrono::Local`
  --> src/filesystem/mod.rs:16:5
   |
16 | use chrono::Local;
   |     ^^^^^^^^^^^^^

warning: unused import: `git2::Repository`
  --> src/filesystem/mod.rs:17:5
   |
17 | use git2::Repository;
   |     ^^^^^^^^^^^^^^^^

warning: unused import: `ignore::overrides::OverrideBuilder`
  --> src/filesystem/mod.rs:18:5
   |
18 | use ignore::overrides::OverrideBuilder;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `ignore::WalkBuilder`
  --> src/filesystem/mod.rs:19:5
   |
19 | use ignore::WalkBuilder;
   |     ^^^^^^^^^^^^^^^^^^^

warning: unused import: `regex::Regex`
  --> src/filesystem/mod.rs:20:5
   |
20 | use regex::Regex;
   |     ^^^^^^^^^^^^

warning: unused import: `std::fs::File`
  --> src/filesystem/mod.rs:21:5
   |
21 | use std::fs::File;
   |     ^^^^^^^^^^^^^

warning: unused import: `std::io::Read`
  --> src/filesystem/mod.rs:22:5
   |
22 | use std::io::Read;
   |     ^^^^^^^^^^^^^

warning: unused import: `std::path::Path`
  --> src/filesystem/mod.rs:23:5
   |
23 | use std::path::Path;
   |     ^^^^^^^^^^^^^^^

warning: unused imports: `path::PathBuf`, `remove_dir`, `remove_file`
  --> src/filesystem/mod.rs:25:10
   |
25 |     fs::{remove_dir, remove_file},
   |          ^^^^^^^^^^  ^^^^^^^^^^^
26 |     path::PathBuf,
   |     ^^^^^^^^^^^^^

warning: unused import: `tracing::error`
  --> src/filesystem/mod.rs:30:5
   |
30 | use tracing::error;
   |     ^^^^^^^^^^^^^^

warning: unused import: `crate::app::Pane`
 --> src/task/search_task.rs:2:5
  |
2 | use crate::app::Pane;
  |     ^^^^^^^^^^^^^^^^

warning: unused import: `crate::app::TaskData`
 --> src/task/search_task.rs:6:5
  |
6 | use crate::app::TaskData;
  |     ^^^^^^^^^^^^^^^^^^^^

warning: unused import: `crate::network::PaneIdentifier`
 --> src/task/search_task.rs:8:5
  |
8 | use crate::network::PaneIdentifier;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `chrono::Local`
  --> src/task/search_task.rs:11:5
   |
11 | use chrono::Local;
   |     ^^^^^^^^^^^^^

warning: unused import: `uuid::Uuid`
  --> src/task/search_task.rs:20:5
   |
20 | use uuid::Uuid;
   |     ^^^^^^^^^^

warning: unreachable expression
   --> src/filesystem/mod.rs:115:36
    |
114 |                         start: todo!(),
    |                                ------- any code following this expression is unreachable
115 |                         completed: todo!(),
    |                                    ^^^^^^^ unreachable expression
    |
    = note: `#[warn(unreachable_code)]` on by default
    = note: this warning originates in the macro `$crate::panic::panic_2015` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused variable: `input`
   --> src/filesystem/mod.rs:109:17
    |
109 |             let input = task.task();
    |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_input`
    |
    = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `e`
   --> src/filesystem/mod.rs:137:17
    |
137 |             Err(e) => {
    |                 ^ help: if this is intentional, prefix it with an underscore: `_e`

warning: unused variable: `search_result_name`
   --> src/task/search_task.rs:110:13
    |
110 |         let search_result_name = store_search_result(&hits2, &self.client_config, &self.input);
    |             ^^^^^^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_search_result_name`

warning: variant is never constructed: `UserActionRequest`
   --> src/app.rs:241:5
    |
241 |     UserActionRequest(UserActionRequest), /* Skip / Abort */
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: variant is never constructed: `FileCopyToExistingFile`
   --> src/app.rs:245:5
    |
245 |     FileCopyToExistingFile, /* Skip / Abort */
    |     ^^^^^^^^^^^^^^^^^^^^^^

warning: enum is never used: `UserActionResponse`
   --> src/app.rs:248:10
    |
248 | pub enum UserActionResponse {
    |          ^^^^^^^^^^^^^^^^^^

warning: associated function is never used: `parent`
   --> src/app.rs:381:12
    |
381 |     pub fn parent(&mut self) -> &mut Option<Pane> {
    |            ^^^^^^

warning: associated function is never used: `new`
  --> src/task/search_task.rs:28:12
   |
28 |     pub fn new(input: SearchDialog, client_config: ClientConfig) -> SearchTask {
   |            ^^^

warning: `rc` (bin "rc") generated 28 warnings
warning: `rc` (bin "rc" test) generated 28 warnings (28 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 1.26s
```

**Comment by SNCPlay42** [comment]
A smaller example:

```rust
fn main() {
    let x = 5;
    loop {}
    println!("{}", x);
}
```

The unused variable warning isn't *wrong* - because the code that uses `x` is never reached, `x` is never used. The problem is that the suggestion fails to realise that although `x` is never used, it is still *named* somewhere else.

The warning in this case should perhaps behave like the one where a variable is assigned to, but never read:

```rust
fn main() {
    let mut x = 5;
    x = 6;
}
```

```
warning: variable `x` is assigned to, but never used
 --> src/lib.rs:2:13
2 |     let mut x = 5;
  |             ^
  |
  = note: `#[warn(unused_variables)]` on by default
  = note: consider using `_x` instead
```

Where the ``consider using `_x` instead`` is not an automatically applicable suggestion.

**Comment by ehuss** [comment]
Thanks for the report!  Closing as a duplicate of #88402.
